onnxruntime/core/graph/graph.cc

```diff
@@ -4046,6 +4070,210 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
   return Status::OK();
 }
 
+static void RenameSubgraphDependentNames(const InlinedHashMap<std::string, std::string>& name_mapping,
+                                         ONNX_NAMESPACE::GraphProto& graph_proto) {
+  for (auto& node : *graph_proto.mutable_node()) {
+    for (auto& attr : *node.mutable_attribute()) {
+      if (attr.has_g()) {
+        RenameSubgraphDependentNames(name_mapping, *attr.mutable_g());
+      }
+
+      for (auto& subgraph : *attr.mutable_graphs()) {
+        RenameSubgraphDependentNames(name_mapping, subgraph);
+      }
+    }
+
+    for (auto& input : *node.mutable_input()) {
+      auto hit = name_mapping.find(input);
+      if (hit != name_mapping.cend()) {
+        input = hit->second;
+      }
+    }
+  }
+}
+
+static void RenameNodeAttributesSubgraphDependentNames(const InlinedHashMap<std::string, std::string>& name_mapping,
+                                                       NodeAttributes& attributes) {
+  for (auto& attribute : attributes) {
+    auto& attr_proto = attribute.second;
+    if (attr_proto.has_g()) {
+      RenameSubgraphDependentNames(name_mapping, *attr_proto.mutable_g());
+    }
+
+    for (auto& subgraph : *attr_proto.mutable_graphs()) {
+      RenameSubgraphDependentNames(name_mapping, subgraph);
+    }
+  }
+}
+
+Status Graph::InlineIfSubgraph(const Graph& graph_to_inline, Node& if_node) {
+  std::string unique_id{if_node.Name()};
+  unique_id.append("_if_").append(graph_to_inline.Name());
+
+  unique_id = GenerateNodeName(unique_id);
+
+  auto make_unique = [&unique_id](const std::string& name) {
+    return unique_id + '_' + name;
+  };
+
+  // Check if the name is an input or implicit input.
+  // These are not renamed.
+  InlinedHashSet<std::string_view> if_all_inputs;
+  const auto if_inputs = if_node.InputDefs();
```

Comment:
Is either (explicit or implicit) inputs ever used down below? Not sure why we would need that, since the renaming is done only for outputs assigned to in some node within the subgraph.

Reply:
The nodes in the graph refer to node inputs, out of scope inputs or the outputs of previously inlined nodes, and the latter  will be new names.